Split, clean up and integrate with tmt ostree.sh

juanje opened the pull request:
* Move the `ostree.sh` script here.
* Remove all the unnecessary parts.
* Split it into different scripts to run as different steps.
* Move part of the setup to a `prepare` phase of `tmt`.
* Move the clean up part to the `finalize` phase of `tmt`.
* Add some minimal smoke test (check that the VM can be accessed via ssh and the OS version).
* Add the **Neptune** app to the image.
* Add support **Centos Stream 8** for **x86_64** and **aarch64**.

lmilbaum commented:
This is a huge PR. Please break it down.

thrix commented on tests/ci/clean_up.sh:
```diff
@@ -0,0 +1,26 @@
+
+cd "$(dirname "${BASH_SOURCE[0]}")"
+
+source /tmp/.env
```
@juanje what environment variables does this host?

juanje replied:
@thrix I split the script in few scripts, but some of them needed vars from others or they are variables shared between different steps. That was the easiest way I found to save the needed variables to be used at the new steps.

Each script add the new variables to the file, so the next one can source and use them.

thrix replied:
makes sense, thx

thrix commented:
@juanje so we can remove the WIP status from here right?

juanje commented:
> @juanje so we can remove the WIP status from here right?

Yes. I've just remove it.

thrix commented:
I rerun the checks to confirm it works on reruns without issues ...

thrix commented on tests/ci/create-commit.sh:
```diff
@@ -0,0 +1,378 @@
+        fi
+        echo -n "."
+
+        # Wait 30 seconds and try again.
```
Maybe 30s is fine, not need to check each 5s. Note that the description is wrong in current version.

juanje replied:
Yes, it's. It was changed at the version I was working for and I didn't realized it was changed.
I'll change it.

thrix commented on tests/ci/create-commit.sh:
```diff
@@ -0,0 +1,378 @@
+
+    # Did the compose finish with success?
+    if [[ $COMPOSE_STATUS != FINISHED ]]; then
+        echo "Something went wrong with the compose. 😢"
```
could we here, `journalctl -n500 -u osbuild-worker\@1` to get more logs?

juanje replied:
I remove the `journalctl` that was running before in background because was noisy and not really useful. When it fails, it usually is more informative the end of the `osbuild-*.log`.

Maybe we could tail that file here. Or both, the log and the journal ooutput :thinking:

thrix commented on tests/ci/create-commit.sh:
```diff
@@ -0,0 +1,378 @@
+sudo composer-cli sources info copr_neptune
+
+# Write a blueprint for ostree image.
+tee "$BLUEPRINT_FILE" > /dev/null << EOF
```
would it make sense to file an issue to the base-enablement team, so they provide the blueprint for installation instead of us hardcoding it into test?

juanje replied:
Yes, it does. This is how this is been done at the `ostree.sh` file for the [upstream](https://github.com/osbuild/osbuild-composer/blob/main/test/cases/ostree.sh#L247) and [downstream](https://github.com/virt-s1/rhel-edge/blob/main/ostree.sh#L304) tests for the `osbuild-composer`, but I prefer the idea of having a file we can track.

Extracting the blueprint is one of the next changes I wanted to do, but it's good idea to have the file sync by the base-enablement team.

BTW, I think we could clean up a bit the blueprint dependencies. I use the blueprint for the aarch64 demo, which has fewer dependencies than the amd64 demo and I remove anther package that we don't really need for the Neptune app.
The sorter the list, the better. Fewer packages to track, smaller image and less time to build.

guy9050 replied:
Just a question - what is this BLUEPRINT used for?

juanje replied:
@guy9050 Sorry, I didn't see the notification.

`$BLUEPRINT` is a file defined [here](https://github.com/osbuild/automotive-ci/blob/main/tests/ci/create-commit.sh#L17) which is used to tell the `composer-cli` the packages and other configs that we need in our image.
Here is where this is used:
https://github.com/osbuild/automotive-ci/blob/main/tests/ci/create-commit.sh#L61

The file has the list of packages that we need to build our image, so it makes sense what @thrix was suggestion about to have it in a separated file and maintain by the ones that know what the image looks like.

thrix commented on tests/ci/files/rhel-8-3-1.json:
```diff
@@ -0,0 +1,26 @@
+{
```
do we need this file here?

thrix replied:
same applies to `rhel-8-4-0.json`

juanje replied:
True. I removed yesterday the RHEL support and I forgot those files. Thanks!

thrix commented:
just few small comments, otherwise nice cleanup, I would go with this version and build on top of it

juanje commented:
> I rerun the checks to confirm it works on reruns without issues ...

The run fails because the virt network was down, but not removed, so and it fails when `setup.sh` tries to destroy the network.
It can fail after failing job or something at the machine. I did something about it at the `clean_up.sh`, but it only catch some cases.
I've added the fix to the `setup.sh`to catch the cases where the previous run wasn't clean enough.

I also changed some of your suggestions. I'll push it now and we'll see if now runs cleanly. And I'll run again, later, to make sure that it handle multiple runs.

juanje commented:
> This is a huge PR. Please break it down.

I splitted a bit the PR in commits, but I couldn't split much more. The first two commits are about moving the script from the external repo to this one. But the functionality doesn't change, just the place where the script is.

The last commit is the actual change:
* Split `ostree.sh` into few scripts/steps.
* Remove the RHEL 8.4 support.
* Add the Neptune packages to the existing `blueprint`.
* Clean up a bit and make sure the CI run fine.

But the other features came from the imported script:
* aarch64 support.
* CentOS Sctream 8 support.

@thrix has already reviewed the new changes and I added some small changes according his comments and suggestions.

lioramilbaum commented on tests/ci/clean_up.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
The following syntax is probably prefered: `#!/usr/bin/env bash`

juanje replied:
I actually prefer the other syntax, so I know that binary is running (not some random `bash` at the `PATH`), and `/bin/` is the common place for `bash` to live in any modern Linux.

I don't mind to change it, but maybe we can agree on some style, so everyone writes the same way. I like this [styleguide](https://google.github.io/styleguide/shellguide.html).
Anyway, I don't think we're going to write a lot of shellscripts... do we?

lioramilbaum replied:
You are probably right. I do not have a strong opinion on this. approving the PR.

lioramilbaum commented on tests/ci/create-commit.sh:
```diff
@@ -0,0 +1,380 @@
+#!/bin/bash
```
See previous comment

lioramilbaum commented on tests/ci/create-commit.sh:
```diff
@@ -0,0 +1,380 @@
+        echo -n "."
+
+        # Wait 30 seconds and try again.
+        sleep 30
```
Personally, I don't like using sleep. This is not a blocker for this PR. Just a general statement.

juanje replied:
This is how the osbuilder team and the qe team do it in their scripts. We need to wait for a while for the osbuild to finish the image. This is the easiest and more common way to do it.
Do you have a better alternative?

lioramilbaum replied:
A suggestion: mage Builder team to provide an API to return if the image is ready or not. If not, return in one of the header the expected amount of time to wait till the image is ready. The idea is inspired by the following (See the Retry-After header) https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

lioramilbaum commented on tests/ci/files/ostree_key:
```diff
@@ -0,0 +1,38 @@
+-----BEGIN OPENSSH PRIVATE KEY-----
```
We are aware that this is a private key in an open source project. Right?

juanje replied:
Yes. Actualy, the InfoSec team detected and ask me about it. We were talking and it wasn't really a risk, as we use them just for the temporary VM, but we agree on trying to generate one new key on the fly each run.

That's one of the few minor changes I wanted to do when the mayor changes get validated.
